motoko/invoice-canister/src/invoice/main.mo

```diff
@@ -0,0 +1,450 @@
+        #err({
+          message = ?"Invoice not found";
+          kind = #NotFound;
+        });
+      };
+      case (?i) {
+        if (i.creator == caller) {
+          return #ok({invoice = i});
+        };
+        // If additional permissions are provided
+        switch (i.permissions) {
+          case (null) {
+            return #err({
+              message = ?"You do not have permission to view this invoice";
+              kind = #NotAuthorized;
+            });
+          };
+          case (?permissions) {
+            let hasPermission = Array.find<Principal>(
+              permissions.canGet,
+              func (x : Principal) : Bool {
+                return x == caller;
+              }
+            );
+            if (Option.isSome(hasPermission)) {
+              return #ok({invoice = i});
+            } else {
+              return #err({
+                message = ?"You do not have permission to view this invoice";
+                kind = #NotAuthorized;
+              });
+            };
+          };
+        };
+        #ok({invoice = i});
+      };
+    };
+  };
+// #endregion
+
+// #region Get Balance
+  public shared ({caller}) func get_balance (args : T.GetBalanceArgs) : async T.GetBalanceResult {
+    let token = args.token;
+    let canisterId = Principal.fromActor(Invoice);
+    switch (token.symbol) {
+      case "ICP" {
+        let defaultAccount = Hex.encode(
+          Blob.toArray(
+            U.getDefaultAccount({
+              canisterId;
+              principal = caller;
+            })
+         )
+        );
+        let balance = await ICP.balance({account = defaultAccount});
+        switch(balance) {
+          case (#err err) {
+            #err({
+              message = ?"Could not get balance";
+              kind = #NotFound;
+            });
+          };
+          case (#ok result){
+            #ok({balance = result.balance});
+          };
+        };
+      };
+      case _ {
+        errInvalidToken;
+      };
+    };
+  };
+// #endregion
+
+// #region Verify Invoice
+  public shared ({caller}) func verify_invoice (args : T.VerifyInvoiceArgs) : async T.VerifyInvoiceResult {
+    let invoice = invoices.get(args.id);
+    let canisterId = Principal.fromActor(Invoice);
+
+    switch invoice {
+      case null{
+        #err({
+          message = ?"Invoice not found";
+          kind = #NotFound;
+        });
+      };
+      case (?i) {
+        // Return if already verified
+        if (i.verifiedAtTime != null) {
+          return #ok(#AlreadyVerified {
+            invoice = i;
+          });
+        };
+        if (i.creator != caller) {
+          switch (i.permissions) {
+            case null {
+              return #err({
+                message = ?"You do not have permission to verify this invoice";
+                kind = #NotAuthorized;
+              });
+            };
+            case (?permissions) {
+              let hasPermission = Array.find<Principal>(
```

Comment:
Suggestion: refactor this into a method, something like`checkPermission(permissions.canVerify, caller) : bool`

Reply:
I'll create a follow-up issue for this